File: src/lib/format/functional-category.ts
import { normalizeStored, toLatinDigits } from "@/lib/normalization/arabic";

export const FUNCTIONAL_CATEGORY_ERROR = 0;
export const FUNCTIONAL_CATEGORY_LABELS = [
  "الفئة الأولى",
  "الفئة الثانية",
  "الفئة الثالثة",
  "الفئة الرابعة",
  "الفئة الخامسة",
] as const;

export type FunctionalCategory = 1 | 2 | 3 | 4 | 5;

/**
 * Converts any Arabic (or numeric) spelling of the functional category into
 * the stored numeric form:
 *
 * - empty / whitespace / null  => null (field is empty)
 * - a recognized category      => 1..5
 * - any other non-empty value  => 0 (error / unknown category)
 *
 * Examples: "1", "٢", "الفئة الأولى", "الأولى", "اولى", "اول", "او" => 1
 * "ثانية", "ثان" => 2, "الثالثة", "لث" => 3, "رابعة", "را" => 4,
 * "خامسة", "مس" => 5.
 */
export function parseFunctionalCategory(value: unknown): FunctionalCategory | 0 | null {
  if (value === null || value === undefined) return null;
  const text = toLatinDigits(String(value)).trim();
  if (!text) return null;

  const numeric = text.match(/^([0-9]+)(?:[.,][0-9]+)?$/);
  if (numeric) {
    const number = Number(numeric[1]);
    return number >= 1 && number <= 5 ? (number as FunctionalCategory) : FUNCTIONAL_CATEGORY_ERROR;
  }

  // Normalize Arabic letters, strip the common filler word(s) and the article.
  const normalized = normalizeStored(text).replace(/الفيه|فيه/g, " ");
  const key = normalized.replace(/\s+/g, "").replace(/^ال/, "").trim();
  if (!key) return null;

  if (/^[0-9]+$/.test(key)) {
    const number = Number(key);
    return number >= 1 && number <= 5 ? (number as FunctionalCategory) : FUNCTIONAL_CATEGORY_ERROR;
  }

  if (key === "او" || key.startsWith("اول")) return 1;
  if (key === "ثان" || key.startsWith("ثان")) return 2;
  if (key === "لث" || key.startsWith("ثالث") || key.startsWith("ثلث")) return 3;
  if (key === "را" || key.startsWith("رابع")) return 4;
  if (key === "مس" || key.startsWith("خامس") || key.startsWith("خمس")) return 5;
  return FUNCTIONAL_CATEGORY_ERROR;
}

/** Display form of a stored category: «الفئة الأولى» … «الفئة الخامسة». */
export function formatFunctionalCategory(value: unknown): string {
  if (value === null || value === undefined || String(value).trim() === "") return "";
  const category = parseFunctionalCategory(value);
  if (category === null) return "";
  if (category === FUNCTIONAL_CATEGORY_ERROR) return "فئة غير معروفة";
  return FUNCTIONAL_CATEGORY_LABELS[category - 1];
}

/**
 * Search helper: returns the numeric category to match in search, or null when
 * the query is not a recognizable category (used to keep full search clean).
 */
export function functionalCategoryQuery(value: unknown): number | null {
  const category = parseFunctionalCategory(value);
  return category !== null && category !== FUNCTIONAL_CATEGORY_ERROR ? category : null;
}

File: backend/ExcelArchive.Domain/Text/FunctionalCategory.cs
using System.Text.RegularExpressions;

namespace ExcelArchive.Domain.Text;

/// <summary>
/// Functional-category parser ported from V1 lib/format/functional-category.ts.
/// Frozen contract: stored numeric form 1..5, 0 = error/unknown, null = empty.
/// </summary>
public static class FunctionalCategory
{
    public const int Error = 0;

    public static readonly string[] Labels =
    [
        "فئة الأولى",
        "فئة الثانية",
        "فئة الثالثة",
        "فئة الرابعة",
        "فئة الخامسة",
    ];

    /// <summary>
    /// Converts any Arabic (or numeric) spelling into 1..5, 0 for unknown
    /// non-empty values, null for empty values. Mirrors V1 exactly, including
    /// the filler-word/article stripping order.
    /// </summary>
    public static int? Parse(object? value)
    {
        if (value is null) return null;
        var text = ArabicNormalizer.ToLatinDigits(value.ToString() ?? "").Trim();
        if (text.Length == 0) return null;

        var numeric = Regex.Match(text, @"^([0-9]+)(?:[.,][0-9]+)?$");
        if (numeric.Success)
        {
            var number = int.Parse(numeric.Groups[1].Value);
            return number is >= 1 and <= 5 ? number : Error;
        }

        var normalized = ArabicNormalizer.NormalizeStored(text).Replace("الفيه", " ").Replace("فيه", " ");
        var key = Regex.Replace(normalized, @"\s+", "").Trim();
        if (key.StartsWith("ال", StringComparison.Ordinal)) key = key[2..];
        key = key.Trim();
        if (key.Length == 0) return null;

        if (Regex.IsMatch(key, "^[0-9]+$"))
        {
            var number = int.Parse(key);
            return number is >= 1 and <= 5 ? number : Error;
        }

        if (key == "او" || key.StartsWith("اول", StringComparison.Ordinal)) return 1;
        if (key.StartsWith("ثان", StringComparison.Ordinal)) return 2;
        if (key == "لث" || key.StartsWith("ثالث", StringComparison.Ordinal) || key.StartsWith("ثلث", StringComparison.Ordinal)) return 3;
        if (key == "را" || key.StartsWith("رابع", StringComparison.Ordinal)) return 4;
        if (key == "مس" || key.StartsWith("خامس", StringComparison.Ordinal) || key.StartsWith("خمس", StringComparison.Ordinal)) return 5;
        return Error;
    }

    /// <summary>Display form of a stored category, mirroring V1.</summary>
    public static string Format(object? value)
    {
        if (value is null || string.IsNullOrWhiteSpace(value.ToString())) return "";
        var category = Parse(value);
        if (category is null) return "";
        if (category == Error) return "فئة غير معروفة";
        return Labels[category.Value - 1];
    }

    /// <summary>Search helper: numeric category or null when unrecognizable.</summary>
    public static int? Query(object? value)
    {
        var category = Parse(value);
        return category is not null && category != Error ? category : null;
    }
}
